File: 프로그래머스/2/42586. 기능개발/기능개발.py
import math

def solution(progresses, speeds):
    max_day = 0
    res = []
    
    for i in range(len(progresses)):
        # 개발은 미리 할 수 있고, 배포는 순서대로
        dev = math.ceil((100 - progresses[i]) / speeds[i]) 
        if i == 0 or max_day < dev:
            max_day = dev
            res.append(1)
        elif max_day >= dev:
            res[-1] += 1
    
    return res

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        Deque<Integer> stack = new ArrayDeque<>();
        int max_day = 0;
        
        for (int i = 0; i < progresses.length; i++) {
            int dev = (int) Math.ceil((double)(100 - progresses[i]) / speeds[i]);
            if (i == 0 || max_day < dev) {
                max_day = dev;
                stack.push(1);
            } else if (max_day >= dev) {
                int tmp = stack.pop();
                stack.push(tmp + 1);
            }
        }
        
        int[] answer = new int[stack.size()];
        int index = 0;
        // 맨 뒤부터 차례대로 꺼내서 배열에 채우기
        while (!stack.isEmpty()) {
            answer[index++] = stack.pollLast();
        }
        
        return answer;
    }
}

